File: android-app/app/src/main/java/com/thoughtinput/capture/data/CapturePayload.kt
package com.thoughtinput.capture.data

import android.os.Build
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.UUID

@Serializable
data class CapturePayload(
    val text: String,
    val timestamp: String,
    @SerialName("source_platform") val sourcePlatform: String = "android",
    @SerialName("client_version") val clientVersion: String = "0.1.0",
    @SerialName("capture_method") val captureMethod: CaptureMethod,
    @SerialName("idempotency_key") val idempotencyKey: String = UUID.randomUUID().toString(),
    @SerialName("device_name") val deviceName: String = (Build.MODEL ?: "unknown")
) {
    @Serializable
    enum class CaptureMethod(val value: String) {
        @SerialName("typed") TYPED("typed"),
        @SerialName("voice") VOICE("voice")
    }

    fun toJson(): String = json.encodeToString(serializer(), this)

    companion object {
        val json = Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
        }

        fun create(
            text: String,
            method: CaptureMethod,
            deviceName: String = (Build.MODEL ?: "unknown")
        ): CapturePayload {
            return CapturePayload(
                text = text,
                timestamp = DateTimeFormatter.ISO_INSTANT.format(Instant.now()),
                captureMethod = method,
                deviceName = deviceName
            )
        }

        fun fromJson(jsonString: String): CapturePayload =
            json.decodeFromString(serializer(), jsonString)
    }
}

File: macos-app/Sources/ThoughtInput/Core/CapturePayload.swift
import Foundation

struct CapturePayload: Codable, Sendable {
    let text: String
    let timestamp: String
    let sourcePlatform: String
    let clientVersion: String
    let captureMethod: CaptureMethod
    let idempotencyKey: String
    let deviceName: String

    enum CaptureMethod: String, Codable, Sendable {
        case typed
        case voice
    }

    enum CodingKeys: String, CodingKey {
        case text
        case timestamp
        case sourcePlatform = "source_platform"
        case clientVersion = "client_version"
        case captureMethod = "capture_method"
        case idempotencyKey = "idempotency_key"
        case deviceName = "device_name"
    }

    static func create(text: String, method: CaptureMethod) -> CapturePayload {
        CapturePayload(
            text: text,
            timestamp: ISO8601DateFormatter().string(from: Date()),
            sourcePlatform: "macos",
            clientVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.1.0",
            captureMethod: method,
            idempotencyKey: UUID().uuidString,
            deviceName: Host.current().localizedName ?? ProcessInfo.processInfo.hostName
        )
    }
}
